### PR TITLE
[DebugInfo][RemoveDIs] Make dropping variable locations explicit

### PR DESCRIPTION
**Pull request by jmorse**
In present-day debug-info, when you delete all instructions, you delete all their debug-info with it because debug-info is stored in instructions. With debug-info stored in DPValue objects however, deleting instructions causes DPValue objects to clump together into a large blob of debug-info that hangs around in the block, as nothing has explicitly deleted it.

To restore this behaviour, scatter calls to dropDbgValues around in places that used to delete chunks of dbg.values, for example during stripDebugInfo and in the code that deletes everything after an Unreachable instruction. DCE is another example.

The tests with --try... added to them are new scenariso where we can now correctly replicate the "normal" debug-info behaviour. Alas, there's no explicit test for the opt -strip-debug option though (in dbg.value mode or DPValue mode).

### CODE REVIEW

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp [hunk 1]
```diff
@@ -5179,6 +5184,15 @@ bool SimplifyCFGOpt::simplifyUnreachable(UnreachableInst *UI) {
 
   bool Changed = false;
 
+  // Ensure that any debug-info records that used to occur after the Unreachable
```
Comment (OCHyams) [inline llvm/lib/Transforms/Utils/SimplifyCFG.cpp:5187]: A few lines up there's an ominous comment `// WARNING: keep in sync with InstCombinerImpl::visitUnreachableInst()!` - does anything over there need changing?

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp [hunk 2]
```diff
@@ -5179,6 +5184,15 @@ bool SimplifyCFGOpt::simplifyUnreachable(UnreachableInst *UI) {
...
+  // are moved to in front of it -- otherwise they'll "dangle" at the end of
+  // the block.
+  BB->flushTerminatorDbgValues();
+
+  // Debug-info records on the unreachable inst itself should be deleted, as
+  // below we delete everything past the final executable instruction.
+  UI->dropDbgValues();
```
Comment (OCHyams) [inline llvm/lib/Transforms/Utils/SimplifyCFG.cpp:5194]: Don't the DPValues on the unreachable come "before" it though?

**Review by OCHyams — APPROVED** [review]
LGTM

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp (hunk 2 above)
Comment (jmorse) [reply llvm/lib/Transforms/Utils/SimplifyCFG.cpp:5194]: They do, and it's arguable that we should keep them, however that's not what the current behaviour with dbg.values is -- the loop immediately below these lines iterates through and deletes any instructions that don't call/branch/signal etc, i.e. anything where it's inevitable that the unreachable will be hit. That includes dbg.values, which get deleted, which is what this addition preserves.

**Comment by jmorse** [comment]
(Ignoring the formatting errors seeing how it's in code I'm not touching)

**Pull request merged**